text/3803-enum-repr-open.md

```diff
@@ -0,0 +1,772 @@
+- Panicking is a common choice for handling the unknown variant (even if one is going to match non-exhaustively, since `match { Ok(Weather::Sunny) => ... }` tend to be quite verbose).
+- It is a bit inefficient (wrapper crates often eagerly check the variants even if the user did not need to check them).
+
+## Binary format serialization/deserialization
+
+When performing zero-copy deserialization of binary data formats, it is often desirable to map a set of bytes to a set of known values, while still handling unknown values. For example, when parsing the Mach-O `LC_BUILD_VERSION` load command, it is desirable to handle the `platform` field as an enum with [certain known values](https://docs.rs/goblin/0.9.3/src/goblin/mach/load_command.rs.html#1336-1347).
+
+This is similar to the case above, but shows that there is value in this feature outside FFI and C interop.
+
+## Greater flexibility for library authors
+
+Winit, a popular GUI creation library in the Rust ecosystem, provides a [`MouseButton`](https://docs.rs/winit/0.30.9/winit/event/enum.MouseButton.html) enum with a few variants like `Left` and `Right`, and a catch-all `Other(u16)` for buttons that do not have a classical semantic meaning.
+
+This is slightly wrong though, since the value `0` actually represents `MouseButton::Left`, but can now be incorrectly represented in the enum as `MouseButton::Other(0)` too. Besides, the enum is slightly larger than it needs to be, which is inefficient.
+
+Winit could have used the `struct` + `const`s pattern, but other mouse button mappings are comparatively rare that the cost of doing this is too high.
+
+The `bitflags` crate might also benefit from marking their enums as `#[repr(_, open)]`.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Imagine you are using a C library that queries the current weather, perhaps from humidity and pressure sensors on the user's device. The library provides the result as a C enumeration.
+
+The author of the library has helpfully provided the following C header:
+
+```c
+// weather.h
+// A service for guessing the current weather.
+
+// The different kinds of weather.
+typedef enum weather : uint8_t { // C23 syntax
+    weather_sunny = 0,
+    weather_windy = 1,
+    weather_rainy = 2,
+} weather_t;
+
+// Guess the current weather.
+weather_t weather_current(void);
+```
+
+This API could be translated to Rust as follows:
+
+```rust
+//! Bindings to a service for guessing the current weather.
+
+/// The different kinds of weather.
+#[repr(u8, open)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum Weather {
+    Sunny = 0,
+    Windy = 1,
+    Rainy = 2,
+}
+
+#[link(name = "weather")]
+// SAFETY: The function signature is correct (return type `weather_t` == `Weather`).
+unsafe extern "C" {
+    // SAFETY: No preconditions for using this API.
+    safe fn weather_current() -> Weather;
+}
```

Comment:
Since it has been noted that future verisons of the C library might add additional consts (and the C library may not consider such an addition to be a breaking change), it would be rather easy to accidentally break memory safety here. I think therefore if you want to yield a Rust enum from a C function, a run-time check of some form should be obligatory (i.e. return `Option<Weather>` or `Result<Weather, UnknownCode>`).